File: app/src/main/java/com/twismart/wallpapershd/ui/main/ListWallpapersFragment.java
package com.twismart.wallpapershd.ui.main;

import android.content.Context;
import android.net.Uri;
import android.os.Bundle;
import android.support.v7.appcompat.BuildConfig;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import com.twismart.wallpapershd.R;
import com.twismart.wallpapershd.data.model.Wallpaper;
import com.twismart.wallpapershd.ui.base.BaseFragment;
import com.twismart.wallpapershd.utils.Constants;

import java.util.ArrayList;

import javax.inject.Inject;


public class ListWallpapersFragment extends BaseFragment implements ListWallpapersContract.View {

    private final String TAG = getClass().getSimpleName();
    private static final String ARG_TYPE_LIST = "typeListWallpapers";
    private OnFragmentInteractionListener mListener;
    private ListWallpapersRecyclerViewAdapter mWallpapersRecyclerViewAdapter;
    private RecyclerView mRecyclerView;
    private String typeListWallpapers;

    @Inject ListWallpapersPresenter<ListWallpapersContract.View> mPresenter;

    public ListWallpapersFragment() {
        // Required empty public constructor
    }

    public static ListWallpapersFragment newInstance(String typeListWallpapers) {
        ListWallpapersFragment fragment = new ListWallpapersFragment();
        Bundle args = new Bundle();
        args.putString(ARG_TYPE_LIST, typeListWallpapers);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            typeListWallpapers = getArguments().getString(ARG_TYPE_LIST);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_list_wallpapers, container, false);

        getActivityComponent().inject(this);

        mRecyclerView = (RecyclerView) v.findViewById(R.id.recycler_wallpapers);
        mRecyclerView.setHasFixedSize(true);
        GridLayoutManager gridLayoutManager = new GridLayoutManager(getContext(), 3);
        mRecyclerView.setLayoutManager(gridLayoutManager);

        mWallpapersRecyclerViewAdapter = new ListWallpapersRecyclerViewAdapter(getContext());
        mRecyclerView.setAdapter(mWallpapersRecyclerViewAdapter);


        Log.d(TAG, "typelist");

        if(typeListWallpapers.equals(Constants.TypeListWallpapers.ALL.value)){
            if(BuildConfig.DEBUG){
                Log.d(TAG, "typelist all");
            }
            else{
                Log.d(TAG, "typelist all nodebug");
            }
            mPresenter.attachView(this);
            mPresenter.getWallpapersList();
            /*
            WallpaperService wallpaperService = WallpaperService.Factory.create();

            Observable<ArrayList<Wallpaper>> wallpapersList = wallpaperService.getWallpapersList("index.php");
            wallpapersList.enqueue(new Callback<ArrayList<Wallpaper>>() {
                @Override
                public void onResponse(Call<ArrayList<Wallpaper>> call, Response<ArrayList<Wallpaper>> response) {
                    Log.d(TAG, "onResponse: code " + response.code());
                    Log.d(TAG, "onResponse: body size " + (response.body().size()));
                    Log.d(TAG, "onResponse: toString " + response.body().toString());
                    mWallpapersRecyclerViewAdapter.setWallpaperList(response.body());
                }
                @Override
                public void onFailure(Call<ArrayList<Wallpaper>> call, Throwable t) {
                    Log.e(TAG, "onFailure: " + t.getMessage());
                }
            });
            */
        }


        return v;
    }

    @Override
    public void setWallpaperList(ArrayList<Wallpaper> wallpaperList) {
        mWallpapersRecyclerViewAdapter.setWallpaperList(wallpaperList);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnFragmentInteractionListener) {
            mListener = (OnFragmentInteractionListener) context;
        } else {
            //throw new RuntimeException(context.toString() + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        mListener = null;
        mPresenter.dettachView();
        super.onDetach();
    }

    public interface OnFragmentInteractionListener {
        void onFragmentInteraction(Uri uri);
    }
}

File: app/src/main/kotlin/com/twismart/wallpapershd/ui/main/ListWallpapersFragment.kt
/*
 * Copyright (C) 2017 Sneyder Angulo.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.twismart.wallpapershd.ui.main

import android.os.Bundle
import android.support.v7.widget.GridLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.twismart.wallpapershd.R
import com.twismart.wallpapershd.data.model.Wallpaper
import com.twismart.wallpapershd.ui.base.BaseFragment
import com.twismart.wallpapershd.utils.Constants
import com.twismart.wallpapershd.utils.debug
import com.twismart.wallpapershd.utils.inflate
import kotlinx.android.synthetic.main.fragment_list_wallpapers.*
import javax.inject.Inject

class ListWallpapersFragment : BaseFragment(), ListWallpapersContract.View {

    companion object {
        private val ARG_TYPE_LIST = "TypeListWallpapers"
        fun newInstance(typeListWallpapers: String): ListWallpapersFragment {
            val fragment = ListWallpapersFragment()
            val args = Bundle()
            args.putString(ARG_TYPE_LIST, typeListWallpapers)
            fragment.arguments = args
            return fragment
        }
    }

    @Inject lateinit var mPresenter: ListWallpapersPresenter<ListWallpapersContract.View>

    private var mWallpapersRecyclerViewAdapter: ListWallpapersRecyclerViewAdapter? = null
    private var mTypeListWallpapers: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            mTypeListWallpapers = arguments.getString(ARG_TYPE_LIST)
        }
    }


    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        activityComponent?.inject(this)

        return context?.inflate(R.layout.fragment_list_wallpapers, container)
    }


    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        mPresenter.attachView(this)

        recyclerViewWallpapers?.apply {
            setHasFixedSize(true)// improve performance
            layoutManager = GridLayoutManager(context, 3)// set grid layout manager with 3 columns
            mWallpapersRecyclerViewAdapter = ListWallpapersRecyclerViewAdapter(context)// create recyclerViewAdapter
            adapter = mWallpapersRecyclerViewAdapter// set recyclerViewAdapter to recyclerView
        }
        refreshContent()

        swipeRefresh?.setOnRefreshListener {
            refreshContent()
        }
    }

    fun refreshContent() {
        debug("refreshContent")
        when (mTypeListWallpapers) {
            Constants.TypeListWallpapers.ALL.value -> {
                mPresenter.getWallpapersList()
            }
            Constants.TypeListWallpapers.MOST_POPULAR.value -> {
                mPresenter.getMostPopularWallpapers()
            }
            Constants.TypeListWallpapers.MY_FAVORITES.value -> {
                mPresenter.getFavoriteWallpapers()
            }
        }
    }

    override fun setWallpaperList(wallpaperList: ArrayList<Wallpaper>) {
        mWallpapersRecyclerViewAdapter?.setWallpaperList(wallpaperList)
        swipeRefresh?.isRefreshing = false
    }

    override fun onDetach() {
        mPresenter.detachView()
        super.onDetach()
    }
}